Skeleton for a Requirements Resolver

willianrampazzo opened the pull request:
This is the initial idea and skeleton for a Requirements Resolver. This code **should not** be reviewed. Feel free to make comments related to the architecture proposed.

**Notes on the current code:**
  - Missing all possible error handling and logs.
  - Works with process and podman, but the podman image should have avocado installed. Feel free to use quay.io/willianr/avocado-82lts.
 - The PackageRequirement class uses the check-installed command and do not install them, so it is safe to run it.
 - Some ugly prints for debug purposes.

**Notes on further development:**
  - While I was implementing the PackageRequirement and creating tasks to be executed by the spawners, I noticed it may make more sense to start a task state machine instead of manually spawn the tasks. This will probably be the approach I'll take.
  - There needs to be a mechanism to fulfill the requirements into a podman image and pass this new image to the state of the state machine responsible to run the tests. The idea here is to create a podman image handler that can be used anywhere before, during or after the test runs, if needed.

Test I'm using for tests:

```python
from avocado import Test


class PassTest(Test):

    """
    Example test that passes.

    :avocado: tags=fast
    :avocado: requirement={"type": "package", "name": "vim-common"}
    :avocado: requirement={"type": "package", "name": "htop"}
    :avocado: requirement={"type": "file", "name": "foo.bar"}
    """

    def test(self):
        """
        A test simply doesn't have to fail in order to pass
        """
```

A run using process spawner and podman spawner:

```
[wrampazz@wrampazz requirements_resolver]$ avocado run --test-runner=nrunner passtest.py 
JOB ID     : ae28f11c10dd81fa05d58c0195eba01850a96b7c
JOB LOG    : /home/wrampazz/avocado/job-results/job-2020-10-23T17.21-ae28f11/job.log
Requirements: [{'type': 'package', 'name': 'vim-common'}, {'type': 'package', 'name': 'htop'}, {'type': 'file', 'name': 'foo.bar'}]
Fulfilling: {'type': 'package', 'name': 'vim-common'}
Fulfilling: {'type': 'package', 'name': 'htop'}
 (1/1) passtest.py:PassTest.test: STARTED
 (1/1) passtest.py:PassTest.test: PASS (0.01 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/wrampazz/avocado/job-results/job-2020-10-23T17.21-ae28f11/results.html
JOB TIME   : 2.89 s

[wrampazz@wrampazz requirements_resolver]$ avocado run --test-runner=nrunner --nrunner-spawner=podman passtest.py 
JOB ID     : bbb429d3fc12e1a58278537e5b49583a78e183c7
JOB LOG    : /home/wrampazz/avocado/job-results/job-2020-10-23T17.22-bbb429d/job.log
Requirements: [{'type': 'package', 'name': 'vim-common'}, {'type': 'package', 'name': 'htop'}, {'type': 'file', 'name': 'foo.bar'}]
Fulfilling: {'type': 'package', 'name': 'vim-common'}
Fulfilling: {'type': 'package', 'name': 'htop'}
 (1/1) passtest.py:PassTest.test: STARTED
 (1/1) passtest.py:PassTest.test: PASS (0.01 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/wrampazz/avocado/job-results/job-2020-10-23T17.22-bbb429d/results.html
JOB TIME   : 4.35 s
```

pevogam commented:
I scanned through the code to get a better grasp of your intended architecture and implementation and I have a few questions:

1) Why is the requirement instance responsible for fulfilling and called fulfiller in some instances? This might lead to circular references and it would probably be better if the requirement instances only hold information on the data of each requirement and methods to update/handle that data. The "fulfilling" part would probably be better of as part of the requirements resolver class itself or subclasses thereof.

2) I see that the fulfillment methods are provided with a spawner which spawns a new task outside of the scope of normal task lifecycles provided in the previous state machine. What could be done to have this fully handled in line with all tasks spawned for actual tests? The task of fulfilling a test requirement could be considered a test precondition and we could for instance try to better integrate the notion of these into the state machine as tasks that are requirements for other tasks. We have something very similar in our Avocado I2N plugin actually, we have "setup tests" which are requirements for other tests and we traverse a graph of such dependencies where the same setup test node could be reused for multiple dependent tests. We could label the tests "tasks" in this case and still consider a coherent way of treating all of them, some simply as requirements of others.

willianrampazzo commented:
> 1. Why is the requirement instance responsible for fulfilling and called fulfiller in some instances? This might lead to circular references and it would probably be better if the requirement instances only hold information on the data of each requirement and methods to update/handle that data. The "fulfilling" part would probably be better of as part of the requirements resolver class itself or subclasses thereof.

I don't think I followed you here. Do you have any line number example? My initial idea is to isolate all the requirements resolution and fulfilling into the Requirements Resolver. What is the reference, or line number, when you say 'requirement instance'?

> 2. I see that the fulfillment methods are provided with a spawner which spawns a new task outside of the scope of normal task lifecycles provided in the previous state machine. What could be done to have this fully handled in line with all tasks spawned for actual tests? 

A way I see to accomplish that is by moving the requirements resolution out of the state machine. Avocado would need to resolve all requirements and add them to the task list before the state machine starts. Considering the state machine's lack of task dependency and the parallel run, this could be a complex task.

> The task of fulfilling a test requirement could be considered a test precondition and we could for instance try to better integrate the notion of these into the state machine as tasks that are requirements for other tasks. We have something very similar in our Avocado I2N plugin actually, we have "setup tests" which are requirements for other tests and we traverse a graph of such dependencies where the same setup test node could be reused for multiple dependent tests. We could label the tests "tasks" in this case and still consider a coherent way of treating all of them, some simply as requirements of others.

The idea of the `triage` node in the status machine is to fulfill all the test preconditions. If any of the test requirements is not available, the test does not run.

I see the state machine as a general use class. We have an instance controlling the test lifecycle. Still, my idea is to instantiate another state machine for controlling the requirements of a test, just like I mentioned in the description of this draft.

Let me know if I misunderstood you in some way and if my comments clarify your doubts.

clebergnu commented on avocado/plugins/spawners/podman.py:
```diff
@@ -29,14 +29,35 @@ def initialize(self):
             help_msg=help_msg,
-            default='fedora:31')
+            # quay.io/willianr/avocado-82lts
+            default='avocado-82lts')
```
You could have used the entire URL here, podman will accept it happily.  But anyway, changing this is annoying, so I sent: https://github.com/avocado-framework/avocado/pull/4284

willianrampazzo replied:
Ah, don't worry about looking at changes here, it was just to make the test easy. As I had a local 82lts image, I forgot to change that before submitting the PR. Thanks for sending #4284, I had it on my to-do list, but never got to it!

clebergnu commented on avocado/plugins/spawners/podman.py:
```diff
@@ -29,14 +29,35 @@ def initialize(self):
+    def podman_bin(self):
+        return self.config.get('spawner.podman.bin')
+
+    async def _copy_to_image(self, source, target, container_id):
```
Nice, there was too much repetition indeed!

clebergnu commented on avocado/plugins/spawners/podman.py:
```diff
@@ -58,6 +79,10 @@ async def spawn_task(self, runtime_task):
         entry_point_args = task.get_command_args()
         entry_point_args.insert(0, "task-run")
         entry_point_args.insert(0, entry_point_cmd)
+        if task.runnable.kind == "avocado-instrumented":
```
I don't think spawners should have to know about task kinds... the information that any spawner need should be generalized.  So, if "avocado-instrumented" tests have a given characteristic that requires an extra handling, these should be generalized to a concept that can be applied to other task/runnable kinds.

This is the idea behind the "core/avocado" requirement.  Other task kinds may have the same requirement, not being able to run unless avocado is present on the environment that will actually run the task.

willianrampazzo replied:
Yep, I agree, as I mentioned in another comment, don't worry about the `podman.py` file, I did it to make things easy on my side. Added it here just for testing purposes, but those changes are not supposed to get merged and don't make much sense as you stated.

clebergnu commented on avocado/plugins/spawners/podman.py:
```diff
@@ -69,13 +94,19 @@ async def spawn_task(self, runtime_task):
             runtime_task.status = msg
             return False
 
+        test_volume = "--volume=AVOCADO:/tmp"
```
I don't understand why this was set here, and then conditionally (because of a task's kind) replaced... care to explain?

clebergnu commented on avocado/plugins/spawners/podman.py:
```diff
@@ -96,19 +127,15 @@ async def spawn_task(self, runtime_task):
-        await proc.wait()
-        if proc.returncode != 0:
-            return False
+        if task.runnable.kind == "avocado-instrumented":
```
IIUC, this block is about changing the location of the test files, so that they can be run from `/tmp`, where they were copied to.  Is that right?

clebergnu commented on avocado/core/requirements/resolver.py:
```diff
@@ -0,0 +1,121 @@
+    This function isolates the Requirements Resolver object into this module.
+
+    :param spawner:
+    :param type: :class:`avocado.plugins.spawner`
```
`:type spawner:`

clebergnu commented on avocado/core/requirements/resolver.py:
```diff
@@ -0,0 +1,121 @@
+    :param spawner:
+    :param type: :class:`avocado.plugins.spawner`
+    :param runtime_task:
+    :param type: :class:`avocado.core.task.runtime.RuntimeTask`
```
`:type runtime_task:`

clebergnu commented:
> This is the initial idea and skeleton for a Requirements Resolver. This code **should not** be reviewed. Feel free to make comments related to the architecture proposed.
> 
> **Notes on the current code:**
> 
> * Missing all possible error handling and logs.
> * Works with process and podman, but the podman image should have avocado installed. Feel free to use quay.io/willianr/avocado-82lts.
> * The PackageRequirement class uses the check-installed command and do not install them, so it is safe to run it.
> * Some ugly prints for debug purposes.
> 
> **Notes on further development:**
> 
> * While I was implementing the PackageRequirement and creating tasks to be executed by the spawners, I noticed it may make more sense to start a task state machine instead of manually spawn the tasks. This will probably be the approach I'll take.

I'm concerned about this, because IMO it will introduce cross state machine dependencies... it could either be that:

1. only once a state machine finishes on state machine "a", their result (all tasks) will be moved to state machine "b"
2. as soon as a task finish their traversal on one state machine "a" they will be moved to state machine "b"

In case 1, it means that tests will not start running as soon as their requirements have been fulfilled.  In case 2, they may have a slight (but negligible) delay. In both cases, though, it seems that it'd increase the complexity of the solution as a whole. 

> * There needs to be a mechanism to fulfill the requirements into a podman image and pass this new image to the state of the state machine responsible to run the tests. The idea here is to create a podman image handler that can be used anywhere before, during or after the test runs, if needed.
> 

OK, this sounds like a good idea. 

> Test I'm using for tests:
> 
> ```python
> from avocado import Test
> 
> 
> class PassTest(Test):
> 
>     """
>     Example test that passes.
> 
>     :avocado: tags=fast
>     :avocado: requirement={"type": "package", "name": "vim-common"}
>     :avocado: requirement={"type": "package", "name": "htop"}
>     :avocado: requirement={"type": "file", "name": "foo.bar"}
>     """
> 
>     def test(self):
>         """
>         A test simply doesn't have to fail in order to pass
>         """
> ```
> 
> A run using process spawner and podman spawner:
> 
> ```
> [wrampazz@wrampazz requirements_resolver]$ avocado run --test-runner=nrunner passtest.py 
> JOB ID     : ae28f11c10dd81fa05d58c0195eba01850a96b7c
> JOB LOG    : /home/wrampazz/avocado/job-results/job-2020-10-23T17.21-ae28f11/job.log
> Requirements: [{'type': 'package', 'name': 'vim-common'}, {'type': 'package', 'name': 'htop'}, {'type': 'file', 'name': 'foo.bar'}]
> Fulfilling: {'type': 'package', 'name': 'vim-common'}
> Fulfilling: {'type': 'package', 'name': 'htop'}
>  (1/1) passtest.py:PassTest.test: STARTED
>  (1/1) passtest.py:PassTest.test: PASS (0.01 s)
> RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
> JOB HTML   : /home/wrampazz/avocado/job-results/job-2020-10-23T17.21-ae28f11/results.html
> JOB TIME   : 2.89 s
> 

So, as it stands, the fulfillment on the process spawner is not working, right? 

> [wrampazz@wrampazz requirements_resolver]$ avocado run --test-runner=nrunner --nrunner-spawner=podman passtest.py 
> JOB ID     : bbb429d3fc12e1a58278537e5b49583a78e183c7
> JOB LOG    : /home/wrampazz/avocado/job-results/job-2020-10-23T17.22-bbb429d/job.log
> Requirements: [{'type': 'package', 'name': 'vim-common'}, {'type': 'package', 'name': 'htop'}, {'type': 'file', 'name': 'foo.bar'}]
> Fulfilling: {'type': 'package', 'name': 'vim-common'}

My expectation when fulfilling any requirement, is to record it in the cache.  It means that if you start with `quay.io/willianr/avocado-82lts:latest` and proceed to fulfill package `vim-common`, a different image will be saved (let's call it `deadbeefdeadbeefdeadbeef`), and the cache/metadata will contain the following information:

_"On environment `deadbeefdeadbeefdeadbeef` you have both environment `quay.io/willianr/avocado-82lts:latest` and package `vim-common`"_

Then the *before* trying to fulfill it on a second run, we could look for those in the cache.

> Fulfilling: {'type': 'package', 'name': 'htop'}

And of coruse the same would apply here, adding *another* image and another package entry.

>  (1/1) passtest.py:PassTest.test: STARTED
>  (1/1) passtest.py:PassTest.test: PASS (0.01 s)
> RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
> JOB HTML   : /home/wrampazz/avocado/job-results/job-2020-10-23T17.22-bbb429d/results.html
> JOB TIME   : 4.35 s
> ```

pevogam commented:
> > 1. Why is the requirement instance responsible for fulfilling and called fulfiller in some instances? This might lead to circular references and it would probably be better if the requirement instances only hold information on the data of each requirement and methods to update/handle that data. The "fulfilling" part would probably be better of as part of the requirements resolver class itself or subclasses thereof.
> 
> I don't think I followed you here. Do you have any line number example? My initial idea is to isolate all the requirements resolution and fulfilling into the Requirements Resolver. What is the reference, or line number, when you say 'requirement instance'?

Sorry I thought we are supposed to send general remarks and review on the choice of implementation and not focus too much on commenting in the code itself. However, now I am realizing that this might have been better in both cases. I was referring to

```python
            fulfiller = type_klass(self._spawner, requirement)
            await fulfiller.fulfill()
```

You are creating a requirement instance and making it fulfill its requirements while a good OOP approach to this would be for it to hold only the the requirement data and possibly checking but leave the fulfillment to a spawner customizer or spawner manager or else.

> > 1. I see that the fulfillment methods are provided with a spawner which spawns a new task outside of the scope of normal task lifecycles provided in the previous state machine. What could be done to have this fully handled in line with all tasks spawned for actual tests?
> 
> A way I see to accomplish that is by moving the requirements resolution out of the state machine. Avocado would need to resolve all requirements and add them to the task list before the state machine starts. Considering the state machine's lack of task dependency and the parallel run, this could be a complex task.

Agreed, the state machine must probably focus only on status tracking, updates, and transitions. I could try to assist you with such dependency resolution implementations in the future if you want. But for now it is still good if at least we agree that tasks shouldn't be initialized and handled both inside the state machine tracking them and outside of it.

> > The task of fulfilling a test requirement could be considered a test precondition and we could for instance try to better integrate the notion of these into the state machine as tasks that are requirements for other tasks. We have something very similar in our Avocado I2N plugin actually, we have "setup tests" which are requirements for other tests and we traverse a graph of such dependencies where the same setup test node could be reused for multiple dependent tests. We could label the tests "tasks" in this case and still consider a coherent way of treating all of them, some simply as requirements of others.
> 
> The idea of the `triage` node in the status machine is to fulfill all the test preconditions. If any of the test requirements is not available, the test does not run.

You mean when the an attempt to fulfill the test requirement is made and fails and the test requirement remains unavailable or all requirements are fullfilled in a batch before any test reaches the triage stage? I can foresee troubles with the second approach since some requirements might be conflicting or in some way incompatible with others. Could requirements be fulfilled on-demand when a test reaches the triage stage?

Another drawback here that I can foresee is that there might be intersections (read: reusable setup) among the requirements that will be duplicated that could be defined for a test. Of course, once could go ahead and split such requirements into smaller nonintersecting modular requirements at the expense of some extra complexity of managing all of them. Note here that by a "requirement" I mean a general requirement which is not necessarily a package (`PackageRequirement`) even though packages too could have common/intersecting dependencies. For me a requirement is associated with an environment (process, container, vm) at a given state and the setup is the action necessary to reach this state i.e. fulfill this requrement. In this case redundancy of setup would result in extra CPU cycles and overall performance costs.

> I see the state machine as a general use class. We have an instance controlling the test lifecycle. Still, my idea is to instantiate another state machine for controlling the requirements of a test, just like I mentioned in the description of this draft.

So for instance if I want to track all tasks where am I supposed to look? If a state machine is about tracking the states of tasks then wouldn't using more than one such instance worsen the tracking itself?

> Let me know if I misunderstood you in some way and if my comments clarify your doubts.

Yes they do, I am just adding some more comments in case they are helpful to you as the person that has the complete picture of what they are planning to implement.

willianrampazzo commented:
Thanks for the comments @pevogam, I'm adding some more :)

> > > 1. Why is the requirement instance responsible for fulfilling and called fulfiller in some instances? This might lead to circular references and it would probably be better if the requirement instances only hold information on the data of each requirement and methods to update/handle that data. The "fulfilling" part would probably be better of as part of the requirements resolver class itself or subclasses thereof.
> > 
> > 
> > I don't think I followed you here. Do you have any line number example? My initial idea is to isolate all the requirements resolution and fulfilling into the Requirements Resolver. What is the reference, or line number, when you say 'requirement instance'?
> 
> Sorry I thought we are supposed to send general remarks and review on the choice of implementation and not focus too much on commenting in the code itself. However, now I am realizing that this might have been better in both cases. I was referring to
> 
> ```python
>             fulfiller = type_klass(self._spawner, requirement)
>             await fulfiller.fulfill()
> ```
> 
> You are creating a requirement instance and making it fulfill its requirements while a good OOP approach to this would be for it to hold only the the requirement data and possibly checking but leave the fulfillment to a spawner customizer or spawner manager or else.
> 

I understood the proposal about the _spawner manager_ handle the fulfillment, but I could not find any reference explaining why it would be a better OOP approach. I looked at several books about OOP and Design Patterns and could not find a reference to something similar. Do you have a reference I can review?

About the proposal I did, in the end, the spawner will handle the fulfillment anyway. Having another object to manage the spawners may make the resulting architecture more complex than it should be. This implementation is based on the BP002, where the Requirements Resolver would serve as a kind of router, sending the requirement information to the correct fulfiller node.

> > > 1. I see that the fulfillment methods are provided with a spawner which spawns a new task outside of the scope of normal task lifecycles provided in the previous state machine. What could be done to have this fully handled in line with all tasks spawned for actual tests?
> > 
> > 
> > A way I see to accomplish that is by moving the requirements resolution out of the state machine. Avocado would need to resolve all requirements and add them to the task list before the state machine starts. Considering the state machine's lack of task dependency and the parallel run, this could be a complex task.
> 
> Agreed, the state machine must probably focus only on status tracking, updates, and transitions. I could try to assist you with such dependency resolution implementations in the future if you want. But for now it is still good if at least we agree that tasks shouldn't be initialized and handled both inside the state machine tracking them and outside of it.
> 

In your opinion, the N(ext)Runner should have just one task state machine running. @clebergnu also added a similar comment about creating other state machines. If we were talking about a full scheduler, I would agree. I was thinking about spawning a new state machine for each task to fulfill its specific requirements. The intent is to reuse the code that already exists to manage a new task. But, if we agree that new tasks should not be created into the state machine, I'll rethink this solution.

Following your proposal to resolve the requirements prior to starting the state machine would work but would need some important changes to the state machine as right now it does not consider task interdependencies. This would also increase the complexity of the architecture, but maybe it is worth it.

> > > The task of fulfilling a test requirement could be considered a test precondition and we could for instance try to better integrate the notion of these into the state machine as tasks that are requirements for other tasks. We have something very similar in our Avocado I2N plugin actually, we have "setup tests" which are requirements for other tests and we traverse a graph of such dependencies where the same setup test node could be reused for multiple dependent tests. We could label the tests "tasks" in this case and still consider a coherent way of treating all of them, some simply as requirements of others.
> > 
> > 
> > The idea of the `triage` node in the status machine is to fulfill all the test preconditions. If any of the test requirements is not available, the test does not run.
> 
> You mean when the an attempt to fulfill the test requirement is made and fails and the test requirement remains unavailable or all requirements are fullfilled in a batch before any test reaches the triage stage? I can foresee troubles with the second approach since some requirements might be conflicting or in some way incompatible with others. Could requirements be fulfilled on-demand when a test reaches the triage stage?
> 

This is the idea, fulfill all the requirements for a task in the `triage` node, on-demand. If while processing a list of requirements, one of them fails, the task is immediately moved to the failed status without the need to check the remaining list of requirements.

> Another drawback here that I can foresee is that there might be intersections (read: reusable setup) among the requirements that will be duplicated that could be defined for a test. Of course, once could go ahead and split such requirements into smaller nonintersecting modular requirements at the expense of some extra complexity of managing all of them. Note here that by a "requirement" I mean a general requirement which is not necessarily a package (`PackageRequirement`) even though packages too could have common/intersecting dependencies. For me a requirement is associated with an environment (process, container, vm) at a given state and the setup is the action necessary to reach this state i.e. fulfill this requrement. In this case redundancy of setup would result in extra CPU cycles and overall performance costs.
> 

Right, the current implementation of the N(ext)Runner relates requirements to tasks. If you have a global requirement, it will be listed for each task (test). In the best scenario, the requirement is fulfilled by the first task that runs and the other tasks just check that the requirement is already there and do not need to be fulfilled. In the worst scenario, two tasks try to fulfill the same requirement at the same time, colliding and failing. This is an open problem that does not have an implementation solution yet.

> > I see the state machine as a general use class. We have an instance controlling the test lifecycle. Still, my idea is to instantiate another state machine for controlling the requirements of a test, just like I mentioned in the description of this draft.
> 
> So for instance if I want to track all tasks where am I supposed to look? If a state machine is about tracking the states of tasks then wouldn't using more than one such instance worsen the tracking itself?
> 

I was thinking about one state machine for the tests and secondary state machines for the requirements. The only available state machine for queries would the test state machine. In case some command queries the state machine about a specific test, if it is still under the `triage` node, the answer would be something like "test is under triage stage with this list of fulfilled and not fulfilled requirements".

> > Let me know if I misunderstood you in some way and if my comments clarify your doubts.
> 
> Yes they do, I am just adding some more comments in case they are helpful to you as the person that has the complete picture of what they are planning to implement.

Again, thanks for the comments, I think they are important and will help us decide which direction to follow for the Requirements Resolver implementation.

willianrampazzo commented:
@pevogam @clebergnu I was sketching the idea about having all requirements resolved to tasks before the task state machine starts, and I came up with the attached diagram, extended from BP003.

With this kind of resolution, a Task will list its requirements as a list of other Task objects. When the Tasks related to tests reach the `Triage` node, they will sit there until all its dependencies Tasks related to requirements reach the `finished` node. If all the requirements Tasks finish with success, the test Task will move to the `Ready` node; otherwise, it will move to the `Finished` node with a fail message.

![Resolution-to-task-requirements-resolver](https://user-images.githubusercontent.com/8462595/98292391-675d1200-1f8b-11eb-813a-72864d020f1d.png)

Again, this is just a sketching of how this could work. Feedbacks are welcome!

richtja commented:
> @pevogam @clebergnu I was sketching the idea about having all requirements resolved to tasks before the task state machine starts, and I came up with the attached diagram, extended from BP003.
> 
> With this kind of resolution, a Task will list its requirements as a list of other Task objects. When the Tasks related to tests reach the `Triage` node, they will sit there until all its dependencies Tasks related to requirements reach the `finished` node. If all the requirements Tasks finish with success, the test Task will move to the `Ready` node; otherwise, it will move to the `Finished` node with a fail message.
> 
> ![Resolution-to-task-requirements-resolver](https://user-images.githubusercontent.com/8462595/98292391-675d1200-1f8b-11eb-813a-72864d020f1d.png)
> 
> Again, this is just a sketching of how this could work. Feedbacks are welcome!

Is there a plan to solve the requirements recursively or we will just deal with the requirements of Tests? Because here I see a problem when two requirement Tasks will their own requirements which could be the same.

willianrampazzo commented:
> > @pevogam @clebergnu I was sketching the idea about having all requirements resolved to tasks before the task state machine starts, and I came up with the attached diagram, extended from BP003.
> > With this kind of resolution, a Task will list its requirements as a list of other Task objects. When the Tasks related to tests reach the `Triage` node, they will sit there until all its dependencies Tasks related to requirements reach the `finished` node. If all the requirements Tasks finish with success, the test Task will move to the `Ready` node; otherwise, it will move to the `Finished` node with a fail message.
> > ![Resolution-to-task-requirements-resolver](https://user-images.githubusercontent.com/8462595/98292391-675d1200-1f8b-11eb-813a-72864d020f1d.png)
> > Again, this is just a sketching of how this could work. Feedbacks are welcome!
> 
> Is there a plan to solve the requirements recursively or we will just deal with the requirements of Tests? Because here I see a problem when two requirement Tasks will their own requirements which could be the same.

IMHO, we should handle test requirements and do not allow a Task requirement to have another requirement. The current architecture already has that kind of dependency in place, where a test reference has a list of requirements. The idea is to iterate over the list of requirements and create the related Task objects.

pevogam commented:
> Again, this is just a sketching of how this could work. Feedbacks are welcome!

Really good visualization! It describes what is in your mind much better in fact. Funnily, I can say that this looks more and more like our own plugin ^^

@richtja: I think indeed solving this recursively is most optimal. However, it is a matter if a very simple check to forbid "reflexive", i.e. circular dependencies so I don't think this could be a wider problem. Now, to take this a step further:

> IMHO, we should handle test requirements and do not allow a Task requirement to have another requirement. The current architecture already has that kind of dependency in place, where a test reference has a list of requirements. The idea is to iterate over the list of requirements and create the related Task objects.

@willianrampazzo I think we could do the opposite since it allows for even greater reuse and recursion and is more natural to read, however we could indeed start with simpler scenarios in order to be able to implement an initial draft. Your proposal here sounds reasonable enough as a first step.

Do you mind taking a look at [this presentation](https://github.com/intra2net/avocado-i2n/blob/master/doc/test_traversal_algorithm.pdf) because I hope it could really give you more ideas, for instance reusing some of my ideas that I described there some years ago. The only difference there is that a requirement is the most general type of requirement I could think of - a vm state and each test is a VT test involving multiple vms and a "fulfillable" state for each vm. We could however imagine here that each test has only one vm and traverse recursively in a similar way.

Just to provide you with some additional background if you are interested to understand better what I mean:
https://github.com/intra2net/avocado-i2n#motivation-and-background
I would appreciate any criticism you have to this approach for dependency resolution as I would like it to be as relatable to what you are doing right now as possible.

pevogam commented:
> > You are creating a requirement instance and making it fulfill its requirements while a good OOP approach to this would be for it to hold only the the requirement data and possibly checking but leave the fulfillment to a spawner customizer or spawner manager or else.
> 
> I understood the proposal about the _spawner manager_ handle the fulfillment, but I could not find any reference explaining why it would be a better OOP approach. I looked at several books about OOP and Design Patterns and could not find a reference to something similar. Do you have a reference I can review?

This was a general advise coming from me since I have often found simplifying class composition help here. However, this is an implementational detail and not too important, there are bigger conceptual decisions to be made first.

> About the proposal I did, in the end, the spawner will handle the fulfillment anyway. Having another object to manage the spawners may make the resulting architecture more complex than it should be. This implementation is based on the BP002, where the Requirements Resolver would serve as a kind of router, sending the requirement information to the correct fulfiller node.

Fair enough and I agree.

> > > > 1. I see that the fulfillment methods are provided with a spawner which spawns a new task outside of the scope of normal task lifecycles provided in the previous state machine. What could be done to have this fully handled in line with all tasks spawned for actual tests?
> > > 
> > > 
> > > A way I see to accomplish that is by moving the requirements resolution out of the state machine. Avocado would need to resolve all requirements and add them to the task list before the state machine starts. Considering the state machine's lack of task dependency and the parallel run, this could be a complex task.
> > 
> > 
> > Agreed, the state machine must probably focus only on status tracking, updates, and transitions. I could try to assist you with such dependency resolution implementations in the future if you want. But for now it is still good if at least we agree that tasks shouldn't be initialized and handled both inside the state machine tracking them and outside of it.
> 
> In your opinion, the N(ext)Runner should have just one task state machine running. @clebergnu also added a similar comment about creating other state machines. If we were talking about a full scheduler, I would agree. I was thinking about spawning a new state machine for each task to fulfill its specific requirements. The intent is to reuse the code that already exists to manage a new task. But, if we agree that new tasks should not be created into the state machine, I'll rethink this solution.

This was another implementational detail, the point in the end is to have as little redundancy and easy tracking as possible.

> Following your proposal to resolve the requirements prior to starting the state machine would work but would need some important changes to the state machine as right now it does not consider task interdependencies. This would also increase the complexity of the architecture, but maybe it is worth it.

No doubt, it won't be easy hehe.

> > Another drawback here that I can foresee is that there might be intersections (read: reusable setup) among the requirements that will be duplicated that could be defined for a test. Of course, once could go ahead and split such requirements into smaller nonintersecting modular requirements at the expense of some extra complexity of managing all of them. Note here that by a "requirement" I mean a general requirement which is not necessarily a package (`PackageRequirement`) even though packages too could have common/intersecting dependencies. For me a requirement is associated with an environment (process, container, vm) at a given state and the setup is the action necessary to reach this state i.e. fulfill this requrement. In this case redundancy of setup would result in extra CPU cycles and overall performance costs.
> 
> Right, the current implementation of the N(ext)Runner relates requirements to tasks. If you have a global requirement, it will be listed for each task (test). In the best scenario, the requirement is fulfilled by the first task that runs and the other tasks just check that the requirement is already there and do not need to be fulfilled. In the worst scenario, two tasks try to fulfill the same requirement at the same time, colliding and failing. This is an open problem that does not have an implementation solution yet.

I guess we could achieve with with an additional "requirement is being currently fulfilled" lock making the task in triage state wait.

> > So for instance if I want to track all tasks where am I supposed to look? If a state machine is about tracking the states of tasks then wouldn't using more than one such instance worsen the tracking itself?
> 
> I was thinking about one state machine for the tests and secondary state machines for the requirements. The only available state machine for queries would the test state machine. In case some command queries the state machine about a specific test, if it is still under the `triage` node, the answer would be something like "test is under triage stage with this list of fulfilled and not fulfilled requirements".

Right if this separation between tests and requirements becomes so distinct, then I guess it will be a necessary complication. I still wonder what stops us from calling the requirements of the types you have currently listed "deployment tests" though. The fact that we are using a state machine again for the requirements already hints at similarities.

The last think I want though is to interrupt your already nearly complete brainstorming. Just throwing a few more ideas on how we saw the same things in the past and have been using them like this for 7-8 years now (since ~2012).

willianrampazzo commented:
@clebergnu @pevogam I think I have enough material for another round of implementation based on the discussion here. I'll try to come up with another PR next week, and we can continue the discussion there.

Thanks again for your valuable comments.

willianrampazzo commented:
Related to #4312.

clebergnu commented:
@willianrampazzo closing this as #4423 supersedes it.
